lib/csv.rb

```diff
@@ -1590,14 +1600,21 @@ def open(filename, mode="r", **options)
       options.delete(:replace)
       options.delete_if {|k, _| /newline\z/.match?(k)}
 
-      begin
-        f = File.open(filename, mode, **file_opts)
-      rescue ArgumentError => e
-        raise unless /needs binmode/.match?(e.message) and mode == "r"
-        mode = "rb"
-        file_opts = {encoding: Encoding.default_external}.merge(file_opts)
-        retry
+      if filename_or_io.is_a?(StringIO)
+        pos = filename_or_io.pos
+        f = StringIO.new(filename_or_io.read)
+        filename_or_io.seek(pos)
```

Comment:
Can we use `StringIO#string` here?

```suggestion
        f = StringIO.new(filename_or_io.string)
```

Reply:
We'd then run into an issue similar to #98, which I confirmed with another test:

```ruby
 def test_foreach_stringio_with_bom
   s = StringIO.new
   s << "\ufeff"  # BOM
   s << @data
   s.rewind
   s.read(3)
   rows = CSV.foreach(s, col_sep: "\t", row_sep: "\r\n").to_a
   assert_equal(@rows, rows)
end
```

Reply:
I guess the most relevant counter-argument is, `CSV.new` also uses `read`, so using `read` here is also a matter of consistency.

Reply:
Good point. Could you also add a test for the BOM case?

How about using `StringIO#set_encoding_by_bom` for it?

```diff
        f = StringIO.new(filename_or_io.string)
        unless f.set_encoding_by_bom
          f.set_encoding(filename_or_io.external_encoding)
        end
```

`CSV.new` uses already opened IO. So it must not ignore the current position. So we must use `read`.
But `CSV.open` opens a new IO. So it must use all contents instead of contents after the current position.
Could you check the current behavior of `CSV.open` with sought real IO?

Reply:
> But CSV.open opens a new IO. So it must use all contents instead of contents after the current position.
You're right, `CSV.open` reads the whole file content. I changed `open` again to use `string` instead of `read`.

> Could you also add a test for the BOM case?
Sure thing. I added it, and together with using the `**file_opts` in the new `StringIO`, it's handled correctly.

> How about using StringIO#set_encoding_by_bom for it?
I couldn't see any effect of doing it. It seems like including the `file_opts` is enough. Or is there a case I'm missing?

Reply:
> I couldn't see any effect of doing it.
 
 Ah, b706d91de8ec640d2918a3b4ff8828bcd6d0f09e is a trick for it.

Reply:
Is there anything you still feel is missing, after I've added `file_opts` with the trick from https://github.com/ruby/csv/commit/b706d91de8ec640d2918a3b4ff8828bcd6d0f09e ?